docs/source/notes/broadcasting.rst

```diff
@@ -86,6 +86,42 @@ For Example::
     >>> (x.add_(y)).size()
     RuntimeError: The expanded size of the tensor (1) must match the existing size (7) at non-singleton dimension 2.
 
+Scatter and gather
+------------------
+
+``gather``, ``scatter_``, ``scatter_add_``, ``scatter``, and ``scatter_add`` also support broadcasting. But the broadcasting
+of these operators are more complicated than most other operators.
+
+When we do ``input.gather(dim, input)``, the following rules are applied:
+
+- The number of dimensions of ``index`` must be equal to or smaller than the number of dimensions of ``input``.
+- If the number of dimensions of ``index`` is smaller than of ``input``, then ``index`` is aligned to the left,
+  that is, extra size 1 dimensions would be appended to the back to make them the same number of dimensions.
+- ``dim`` could be negative, if this is the case, then the ``dim`` is wrapped according to the original dimensions
+  if ``index``, not the broadcasted dimension. For example, if ``input`` is 8-dimensional and ``index`` is 4-dimensional,
+  then ``dim=-1`` is the same as ``dim=3``.
```

Comment:
I find this mildly shocking.

Let's consider the `torch.gather` example.

```
out[i][j][k] = input[index[i][j][k]][j][k]  # if dim == 0
out[i][j][k] = input[i][index[i][j][k]][k]  # if dim == 1
out[i][j][k] = input[i][j][index[i][j][k]]  # if dim == 2
```

When index is broadcasted from 2D to 3D, we end up with equations like this:

```
out[i][j][k] = input[index[i][j]][j][k]  # if dim == 0
out[i][j][k] = input[i][index[i][j]][k]  # if dim == 1
out[i][j][k] = input[i][j][index[i][j]]  # if dim == 2
```

Now, let us consider the negative dimension cases. With your rule, I believe we get:

```
out[i][j][k] = input[index[i][j]][j][k]  # if dim == -2
out[i][j][k] = input[i][index[i][j]][k]  # if dim == -1
```

I find this incredibly shocking. The semantics of gather is "gather elements along the axis specified by dim." If I say -1, I expect this to refer to the axis of the INPUT I am gathering from, not the indexing tensor!